stdlib/public/core/UnsafeBufferPointer.swift.gyb

```diff
@@ -193,11 +219,19 @@ extension Unsafe${Mutable}BufferPointer: ${Mutable}Collection, RandomAccessColle
     // optimizer is not capable of creating partial specializations yet.
     // NOTE: Range checks are not performed here, because it is done later by
     // the subscript function.
-    let l = limit - i
+    // NOTE: Unchecked math because unsafe pointers are not required to
+    // behave sensibly if you misuse their indices. We validate in
+    // debug mode.
+    let result = limit.subtractingReportingOverflow(i)
+    _debugPrecondition(!result.overflow)
+    let l = result.partialValue
+
     if n > 0 ? l >= 0 && l < n : l <= 0 && n < l {
       return nil
     }
-    return i + n
+
+    // No need to check here: n is within bounds per the check above.
+    return i &+ n
```

Comment:
I don't really see how the checks above have anything to do with whether or not the addition can overflow here.

I think the point isn't that the addition is guaranteed not to overflow at this point (that's simply not true -- the comment is misleading!), but that we decided we don't care if it does.

```suggestion
    let result = i.addingReportingOverflow(n)
    _debugPrecondition(!result.overflow)
    return result.partialValue
```

Reply:
In fact, the use of the term "Unchecked math" in the comments is just not correct, and this point highlights that.

Unlike C, signed integer overflow in Swift is well-defined. Always. So what does it do?

- For general arithmetic, the `+` operator traps on overflow. Because it's the safest default when you have no context about how the integer will be used.
- For code with more context, the `&+` operator allows wrapping on overflow instead. It is important to note that this is well-defined. It is categorically not correct to describe this as being unsafe, unchecked, undefined, etc.

The distinction is between wrapping vs. trapping, not 'checked vs. unchecked'. `&+` only offers more performance because it happens to match what most hardware platforms do by default anyway, so we can avoid some branches. For more exotic hardware where that is not the case, extra checks _would_ need to be inserted to ensure that they match the `&+` operator's well-defined semantics.